xla/pjrt/gpu/se_gpu_pjrt_client.cc

```diff
@@ -393,6 +408,249 @@ absl::StatusOr<std::unique_ptr<Communicator>> CreateTransferCommunicator(
+  };
+}
+
+// Create a `PreparedReceive` object bundling together state needed to perform a
+// receive.
+absl::StatusOr<PreparedReceive> PrepareReceive(
+    StreamExecutorGpuClient* client, gpu::GpuCollectives* gpu_collectives,
+    se::Stream* stream, PjRtDevice* device, PjRtMemorySpace* memory_space,
+    PjRtGlobalDeviceId src_global_device_id, CrossHostTransferKey transfer_key,
+    Shape shape, gpu::AcquiredCliquesMap& acquired_cliques_map) {
+  // Form the GPU clique key.
+  // TODO(asrao, mwhittaker): Supply correct incarnations when creating the
+  // clique key.
+  gpu::GpuCliqueKey clique_key = gpu::GpuCliqueKey(
+      /*devices=*/{GlobalDeviceId(src_global_device_id.value()),
+                   GlobalDeviceId(device->global_device_id().value())},
+      /*num_local_participants=*/1);
+
+  // Get the clique and communicator for the receive.
+  TF_ASSIGN_OR_RETURN(
+      AcquiredCliqueAndCommunicator clique_and_communicator,
+      AcquireCliqueAndCommunicator(client, gpu_collectives, clique_key,
+                                   acquired_cliques_map, RankId(1), stream));
+
+  // Create a definition event to track when the receive is done.
+  tsl::RCReference<PjRtStreamExecutorDeviceEvent> definition_event =
+      tsl::MakeRef<PjRtStreamExecutorDeviceEvent>(
+          BufferSequencingEvent::Create(client->thread_pool()));
+
+  // Allocate an uninitialized buffer. The buffer will be populated with data
+  // received from the sending process.
+  TF_ASSIGN_OR_RETURN(
+      Shape on_device_shape,
+      client->MakeDefaultShapeForMemorySpace(
+          memory_space, shape, shape.has_layout() ? &shape.layout() : nullptr));
+  TF_ASSIGN_OR_RETURN(
+      size_t on_device_bytes_count,
+      client->GetOnDeviceBytesCount(memory_space, on_device_shape));
+  TF_ASSIGN_OR_RETURN(
+      tsl::RCReference<CommonPjRtRawBuffer> raw_buffer,
+      client->AllocateRawBuffer(memory_space, on_device_bytes_count,
+                                /*retry_on_oom=*/true,
+                                /*allocate_after=*/{}));
+  TF_ASSIGN_OR_RETURN(
+      std::unique_ptr<PjRtBuffer> buffer,
+      client->DefineBuffer(on_device_shape, raw_buffer, {definition_event},
+                           /*raw_buffer_is_mutable=*/true));
+  definition_event->AndThen([raw_buffer]() {});
+
+  return PreparedReceive{
+      /*client=*/client,
+      /*clique_key=*/std::move(clique_key),
+      /*buffer=*/std::move(buffer),
+      /*raw_buffer=*/std::move(raw_buffer),
+      /*definition_event=*/std::move(definition_event),
+      /*clique_and_communicator=*/std::move(clique_and_communicator),
+  };
+}
+
+// Groups Prepared{Send,Receive} structs that have a `.clique_key` attribute.
+template <typename PreparedT>
+static absl::flat_hash_map<gpu::GpuCliqueKey, std::vector<PreparedT>>
+GroupByCliqueKey(std::vector<PreparedT>&& prepareds) {
+  absl::flat_hash_map<gpu::GpuCliqueKey, std::vector<PreparedT>> grouped;
+  grouped.reserve(prepareds.size());
```

Comment:
Am I understanding correctly that `grouped` can be smaller than `prepared.size()`? If so, we can skip the `reserve` call?